File: utils/estimationLogic.ts

import type { BatchData, SimilarDay } from '../types';
import { timeToMinutes, minutesToTime, roundToNearest5 } from './time';

export interface EstimationResult {
  pointEstimateMinutes: number;
  volumeAdjustment: number;
  baseTimeMinutes: number;
  bestCandidate: SimilarDay | null;
  strategyUsed: string;
  finalPrediction: {
    start: string;
    end: string;
  } | null;
  // Fields for Range Model
  smartDetails?: {
    neighbors: SimilarDay[];
    averageDelay: number;
    confidenceScore: string; // "High", "Medium", "Low"
    dynamicRangeWidth: number;
    safetyBias: number;
    specialAdjustments?: string[]; // To explain the new rules
    skewInfo?: {
        skewVal: number;
        description: string;
        leftOffset: number;
        rightOffset: number;
    }
  }
}

/**
 * Helper to calculate Median
 */
const calculateMedian = (values: number[]): number => {
    if (values.length === 0) return 0;
    const sorted = [...values].sort((a, b) => a - b);
    const mid = Math.floor(sorted.length / 2);
    if (sorted.length % 2 !== 0) {
        return sorted[mid];
    }
    return (sorted[mid - 1] + sorted[mid]) / 2;
};



/**
 * DEPRECATED: Old "Hernan" Heuristic Logic.
 */
export const findBestCandidateLegacy = (target: BatchData, history: BatchData[]): { candidate: SimilarDay | null, strategy: string } => {
  const pool = history.filter(h => h.id !== target.id && (h.debitos + h.creditos) > 0);

  const sortByVolume = (candidates: BatchData[]) => {
    return candidates.map(job => ({
      ...job,
      debitDifference: Math.abs(job.debitos - target.debitos),
    })).sort((a, b) => a.debitDifference - b.debitDifference);
  };

  const strictMatches = pool.filter(h => 
    h.especial.toLowerCase() === target.especial.toLowerCase() && 
    h.dia === target.dia
  );
  
  if (strictMatches.length > 0) {
    return { candidate: sortByVolume(strictMatches)[0], strategy: 'Legado: Estricta' };
  }

  return { candidate: null, strategy: 'Sin Datos' };
};

export const calculateEstimationLegacy = (targetJob: BatchData, historicalData: BatchData[], manualExtraTime: number = 0): EstimationResult => {
    const { candidate: bestCandidate, strategy } = findBestCandidateLegacy(targetJob, historicalData);

    if (!bestCandidate) {
        return {
            pointEstimateMinutes: 0,
            volumeAdjustment: 0,
            baseTimeMinutes: 0,
            bestCandidate: null,
            strategyUsed: 'None',
            finalPrediction: null
        };
    }

    const debitDiff = targetJob.debitos - bestCandidate.debitos;
    const rawAdjustment = (debitDiff / 100000) * 20.0;
    const roundedAdjustment = Math.round(rawAdjustment / 5) * 5;

    const baseTime = timeToMinutes(bestCandidate.horaSinDemora);
    const prelimMinutes = baseTime + roundedAdjustment + manualExtraTime;
    const centerPoint = roundToNearest5(prelimMinutes);

    const finalStartMinutes = centerPoint - 15;
    const finalEndMinutes = centerPoint + 15;
    
    return {
        pointEstimateMinutes: centerPoint,
        volumeAdjustment: roundedAdjustment,
        baseTimeMinutes: baseTime,
        bestCandidate,
        strategyUsed: strategy,
        finalPrediction: {
            start: minutesToTime(finalStartMinutes),
            end: minutesToTime(finalEndMinutes),
        }
    };
};

/**
 * NEW LOGIC: Concepto de Rango (Range Concept) v6.0 - Asymmetric Precision
 * Keeps 30m total width, but shifts the window based on distribution skew.
 */
export const calculateSmartEstimation = (targetJob: BatchData, historicalData: BatchData[], manualExtraTime: number = 0): EstimationResult => {
    
    const specialRulesApplied: string[] = [];
    const targetSpecialLower = targetJob.especial.toLowerCase();

    // 1. Scoring Candidates
    const pool = historicalData.filter(h => h.id !== targetJob.id);
    
    const scoredCandidates = pool.map(job => {
        let score = 0;
        if (job.especial.toLowerCase() === targetSpecialLower) score += 50;
        
        const volDiff = Math.abs(job.debitos - targetJob.debitos);
        if (volDiff < 50000) score += 40;
        else if (volDiff < 100000) score += 20;
        else if (volDiff < 200000) score += 5;
        else score -= 10;

        if (job.dia === targetJob.dia) score += 10;

        return { ...job, score, debitDifference: volDiff, includedInCalc: true };
    }).sort((a, b) => b.score - a.score);

    if (scoredCandidates.length === 0) return calculateEstimationLegacy(targetJob, historicalData, manualExtraTime);

    // --- DYNAMIC CLUSTERING ---
    const bestScore = scoredCandidates[0].score;
    let selectedNeighbors: typeof scoredCandidates = [];

    const relativeThreshold = bestScore * 0.75;
    selectedNeighbors = scoredCandidates.filter(c => c.score >= relativeThreshold || c.score > 80);

    if (selectedNeighbors.length < 3) {
        selectedNeighbors = scoredCandidates.slice(0, 3);
        specialRulesApplied.push('Cluster Extendido');
    }
    if (selectedNeighbors.length > 12) {
        selectedNeighbors = selectedNeighbors.slice(0, 12);
    }

    // 2. Calculate Volatility & Median Drift
    const neighborsWithDrift = selectedNeighbors.map(n => {
        const real = timeToMinutes(n.horaReal);
        const noDelay = timeToMinutes(n.horaSinDemora);
        let d = real - noDelay;
        if (d < -1000) d += 1440; 
        return { ...n, drift: d };
    });

    const drifts = neighborsWithDrift.map(n => n.drift);
    const meanDrift = drifts.reduce((a, b) => a + b, 0) / drifts.length;
    const variance = drifts.reduce((a, b) => a + Math.pow(b - meanDrift, 2), 0) / drifts.length;
    const stdDev = Math.sqrt(variance);

    // Filter Outliers for Base Calculation
    const validDrifts: number[] = [];
    const processedNeighbors = neighborsWithDrift.map(n => {
        const threshold = Math.max(stdDev * 1.5, 15); 
        const isOutlier = Math.abs(n.drift - meanDrift) > threshold;
        if (!isOutlier) {
            validDrifts.push(n.drift);
        }
        return { ...n, includedInCalc: !isOutlier };
    });

    let predictedDrift = calculateMedian(validDrifts);
    
    // Damping logic
    if (predictedDrift > 15 && stdDev > 20) {
        predictedDrift = predictedDrift * 0.7; 
        specialRulesApplied.push('Filtro Ruido');
    }

    // 3. Base Time
    let targetNoDelay = 0;
    if (!targetJob.horaSinDemora || targetJob.horaSinDemora === 'N/A' || targetJob.horaSinDemora === '') {
         const validNeighbors = processedNeighbors.filter(n => n.includedInCalc);
         const sumBase = validNeighbors.reduce((acc, n) => acc + timeToMinutes(n.horaSinDemora), 0);
         targetNoDelay = sumBase / (validNeighbors.length || 1);
    } else {
         targetNoDelay = timeToMinutes(targetJob.horaSinDemora);
    }
    
    // 4. Volume Adjustment
    const validCount = validDrifts.length;
    const avgNeighborDebits = processedNeighbors.filter(n => n.includedInCalc).reduce((sum, n) => sum + n.debitos, 0) / (validCount || 1);
    let volumeDiff = targetJob.debitos - avgNeighborDebits;
    let volumeAdjustment = (volumeDiff / 100000) * 8; 

    const efficiencyBias = -5;
    
    // 5. Center Point Calculation
    let predictedMinutes = targetNoDelay + predictedDrift + volumeAdjustment + efficiencyBias + manualExtraTime;
    const centerPoint = roundToNearest5(predictedMinutes);

    // --- ASYMMETRIC 30 MIN INTERVAL ---
    // Analyze Skew: Compare Mean vs Median of Valid Drifts.
    // If Mean > Median, it implies a "Right Tail" (Late outliers). We shift the 30m window to the right.
    const validMeanDrift = validDrifts.length > 0 ? validDrifts.reduce((a, b) => a + b, 0) / validDrifts.length : 0;
    const skew = validMeanDrift - predictedDrift;

    // Base: Symmetric 30m (-15 / +15)
    let rangeStartOffset = 15;
    let rangeEndOffset = 15;
    let skewDescription = "Simétrico";

    // Skew Logic: Shift the window while keeping Total Width = 30
    if (skew > 2) {
        // Strong tendency to be late -> Shift Window Right relative to Center
        // Example: Center is 02:00. Instead of 01:45-02:15, use 01:50-02:20 (-10/+20)
        rangeStartOffset = 10;
        rangeEndOffset = 20;
        skewDescription = "Asimétrico (Cola Derecha)";
        specialRulesApplied.push('Asimetría: +Late Risk');
    } else if (skew < -2) {
        // Tendency to be early/stable -> Shift Window Left relative to Center
        rangeStartOffset = 20;
        rangeEndOffset = 10;
        skewDescription = "Asimétrico (Cola Izquierda)";
        specialRulesApplied.push('Asimetría: +Early Bias');
    }

    const totalWidth = rangeEndOffset + rangeStartOffset; // Always 30

    const rangeStart = centerPoint - rangeStartOffset;
    const rangeEnd = centerPoint + rangeEndOffset;
    
    specialRulesApplied.push('Rango Fijo (30m)');

    return {
        pointEstimateMinutes: centerPoint,
        volumeAdjustment: Math.round(volumeAdjustment),
        baseTimeMinutes: Math.round(targetNoDelay),
        bestCandidate: processedNeighbors[0],
        strategyUsed: `Eficiencia Asimétrica`,
        finalPrediction: {
            start: minutesToTime(rangeStart),
            end: minutesToTime(rangeEnd),
        },
        smartDetails: {
            neighbors: processedNeighbors,
            averageDelay: Math.round(predictedDrift),
            confidenceScore: stdDev < 15 ? 'Alta' : 'Moderada',
            dynamicRangeWidth: totalWidth,
            safetyBias: 0, // Not used in this logic
            specialAdjustments: specialRulesApplied,
            skewInfo: {
                skewVal: parseFloat(skew.toFixed(2)),
                description: skewDescription,
                leftOffset: rangeStartOffset,
                rightOffset: rangeEndOffset
            }
        }
    };
};

export const findSimilarDays = (target: BatchData, history: BatchData[]): SimilarDay[] => {
   return history
    .filter(job => job.id !== target.id && job.especial === target.especial)
    .map(job => ({...job, debitDifference: Math.abs(job.debitos - target.debitos)}))
    .sort((a, b) => a.debitDifference - b.debitDifference)
    .slice(0, 5); 
}

export const checkAccuracy = (prediction: { start: string, end: string }, actualTime: string): boolean => {
    if (!actualTime || actualTime === 'N/A' || actualTime === '') return false;

    const start = timeToMinutes(prediction.start);
    const end = timeToMinutes(prediction.end);
    const actual = timeToMinutes(actualTime);
    
    const GRACE_PERIOD = 5; // AI tolerance
    
    let adjStart = start;
    let adjEnd = end;
    let adjActual = actual;

    if (adjEnd < adjStart) {
        adjEnd += 1440;
    }
    if (adjActual < adjStart - 180) { 
         adjActual += 1440;
    }
    else if (adjActual > adjEnd + 180 && adjStart < 300) { 
        adjActual -= 1440;
    }

    return adjActual >= (adjStart - GRACE_PERIOD) && adjActual <= (adjEnd + GRACE_PERIOD);
};

export const checkManualAccuracy = (prediction: { start: string, end: string }, actualTime: string): boolean => {
    if (!actualTime || actualTime === 'N/A' || actualTime === '') return false;

    const start = timeToMinutes(prediction.start);
    const end = timeToMinutes(prediction.end);
    const actual = timeToMinutes(actualTime);
    
    const MANUAL_GRACE_PERIOD = 0; // No tolerance for manual - must be exact
    
    let adjStart = start;
    let adjEnd = end;
    let adjActual = actual;

    if (adjEnd < adjStart) {
        adjEnd += 1440;
    }
    if (adjActual < adjStart - 180) { 
         adjActual += 1440;
    }
    else if (adjActual > adjEnd + 180 && adjStart < 300) { 
        adjActual -= 1440;
    }

    return adjActual >= (adjStart - MANUAL_GRACE_PERIOD) && adjActual <= (adjEnd + MANUAL_GRACE_PERIOD);
};

export const checkAIWithEarlyAccuracy = (prediction: { start: string, end: string }, actualTime: string): boolean => {
    if (!actualTime || actualTime === 'N/A' || actualTime === '') return false;

    const start = timeToMinutes(prediction.start);
    const end = timeToMinutes(prediction.end);
    const actual = timeToMinutes(actualTime);
    
    const GRACE_PERIOD = 5; // Tolerance only for lower limit
    
    let adjStart = start;
    let adjEnd = end;
    let adjActual = actual;

    if (adjEnd < adjStart) {
        adjEnd += 1440;
    }
    if (adjActual < adjStart - 180) { 
         adjActual += 1440;
    }
    else if (adjActual > adjEnd + 180 && adjStart < 300) { 
        adjActual -= 1440;
    }

    // Strict on upper limit (no positive deviations allowed)
    // Tolerant on lower limit (±5 min tolerance for early cases)
    return adjActual >= (adjStart - GRACE_PERIOD) && adjActual <= adjEnd;
};

export const parseManualRange = (rangeStr: string): { start: string, end: string } | null => {
    if (!rangeStr || rangeStr === 'N/A') return null;
    const parts = rangeStr.split('-').map(s => s.trim());
    if (parts.length !== 2) return null;
    return { start: parts[0], end: parts[1] };
};

File: python_agent/estimation_logic.py

import pandas as pd
import numpy as np
from typing import List, Dict, Any

# Utilidades de tiempo
def time_to_minutes(time_str: str) -> int:
    parts = time_str.split(":")
    hours = int(parts[0])
    minutes = int(parts[1])
    return hours * 60 + minutes

def minutes_to_time(total_minutes: int) -> str:
    hours = total_minutes // 60
    minutes = round(total_minutes % 60)
    return f"{hours:02d}:{minutes:02d}"

def round_to_nearest_5(minutes: float) -> int:
    return int(round(minutes / 5) * 5)

def calculate_median(values: List[float]) -> float:
    if not values:
        return 0.0
    sorted_vals = sorted(values)
    mid = len(sorted_vals) // 2
    if len(sorted_vals) % 2 != 0:
        return sorted_vals[mid]
    return (sorted_vals[mid - 1] + sorted_vals[mid]) / 2

# Nodo IA principal: estimación inteligente
def calculate_smart_estimation(target_job: Dict[str, Any], historical_data: List[Dict[str, Any]], manual_extra_time: float = 0) -> Dict[str, Any]:
    special_rules_applied = []
    target_special_lower = target_job['especial'].lower()

    # 1. Scoring Candidates
    pool = [h for h in historical_data if h['id'] != target_job['id']]
    scored_candidates = []
    for job in pool:
        score = 0
        if job['especial'].lower() == target_special_lower:
            score += 50
        vol_diff = abs(job['debitos'] - target_job['debitos'])
        if vol_diff < 50000:
            score += 40
        elif vol_diff < 100000:
            score += 20
        elif vol_diff < 200000:
            score += 5
        else:
            score -= 10
        if job['dia'] == target_job['dia']:
            score += 10
        job_copy = job.copy()
        job_copy['score'] = score
        job_copy['debitDifference'] = vol_diff
        job_copy['includedInCalc'] = True
        scored_candidates.append(job_copy)
    scored_candidates.sort(key=lambda x: -x['score'])

    if not scored_candidates:
        return {"error": "No candidates found"}

    best_score = scored_candidates[0]['score']
    relative_threshold = best_score * 0.75
    selected_neighbors = [c for c in scored_candidates if c['score'] >= relative_threshold or c['score'] > 80]
    if len(selected_neighbors) < 3:
        selected_neighbors = scored_candidates[:3]
        special_rules_applied.append('Cluster Extendido')
    if len(selected_neighbors) > 12:
        selected_neighbors = selected_neighbors[:12]

    # 2. Calcular drift y filtrar outliers
    neighbors_with_drift = []
    for n in selected_neighbors:
        real = time_to_minutes(n['horaReal'])
        no_delay = time_to_minutes(n['horaSinDemora'])
        d = real - no_delay
        if d < -1000:
            d += 1440
        n_copy = n.copy()
        n_copy['drift'] = d
        neighbors_with_drift.append(n_copy)
    drifts = [n['drift'] for n in neighbors_with_drift]
    mean_drift = np.mean(drifts) if drifts else 0
    std_dev = np.std(drifts) if drifts else 0
    valid_drifts = []
    processed_neighbors = []
    for n in neighbors_with_drift:
        threshold = max(std_dev * 1.5, 15)
        is_outlier = abs(n['drift'] - mean_drift) > threshold
        n['includedInCalc'] = not is_outlier
        if not is_outlier:
            valid_drifts.append(n['drift'])
        processed_neighbors.append(n)
    predicted_drift = calculate_median(valid_drifts)
    if predicted_drift > 15 and std_dev > 20:
        predicted_drift = predicted_drift * 0.7
        special_rules_applied.append('Filtro Ruido')

    # 3. Base Time
    if not target_job.get('horaSinDemora') or target_job['horaSinDemora'] in ['N/A', '']:
        valid_neighbors = [n for n in processed_neighbors if n['includedInCalc']]
        sum_base = sum(time_to_minutes(n['horaSinDemora']) for n in valid_neighbors)
        target_no_delay = sum_base / (len(valid_neighbors) or 1)
    else:
        target_no_delay = time_to_minutes(target_job['horaSinDemora'])

    # 4. Volume Adjustment
    valid_count = len(valid_drifts)
    avg_neighbor_debits = sum(n['debitos'] for n in processed_neighbors if n['includedInCalc']) / (valid_count or 1)
    volume_diff = target_job['debitos'] - avg_neighbor_debits
    volume_adjustment = (volume_diff / 100000) * 8
    efficiency_bias = -5

    # 5. Center Point Calculation
    predicted_minutes = target_no_delay + predicted_drift + volume_adjustment + efficiency_bias + manual_extra_time
    center_point = round_to_nearest_5(predicted_minutes)

    # Skew
    valid_mean_drift = np.mean(valid_drifts) if valid_drifts else 0
    skew = valid_mean_drift - predicted_drift
    range_start_offset = 15
    range_end_offset = 15
    skew_description = "Simétrico"
    if skew > 2:
        range_start_offset = 10
        range_end_offset = 20
        skew_description = "Asimétrico (Cola Derecha)"
        special_rules_applied.append('Asimetría: +Late Risk')
    elif skew < -2:
        range_start_offset = 20
        range_end_offset = 10
        skew_description = "Asimétrico (Cola Izquierda)"
        special_rules_applied.append('Asimetría: +Early Bias')
    total_width = range_end_offset + range_start_offset
    range_start = center_point - range_start_offset
    range_end = center_point + range_end_offset
    special_rules_applied.append('Rango Fijo (30m)')

    return {
        "pointEstimateMinutes": center_point,
        "volumeAdjustment": round(volume_adjustment),
        "baseTimeMinutes": round(target_no_delay),
        "bestCandidate": processed_neighbors[0] if processed_neighbors else None,
        "strategyUsed": "Eficiencia Asimétrica",
        "finalPrediction": {
            "start": minutes_to_time(range_start),
            "end": minutes_to_time(range_end)
        },
        "smartDetails": {
            "neighbors": processed_neighbors,
            "averageDelay": round(predicted_drift),
            "confidenceScore": 'Alta' if std_dev < 15 else 'Moderada',
            "dynamicRangeWidth": total_width,
            "safetyBias": 0,
            "specialAdjustments": special_rules_applied,
            "skewInfo": {
                "skewVal": round(skew, 2),
                "description": skew_description,
                "leftOffset": range_start_offset,
                "rightOffset": range_end_offset
            }
        }
    }
